tests/e2e/browse-bills.spec.ts

```diff
@@ -0,0 +1,272 @@
+    await billpage.goto()
+
+    const searchTerm = billpage.searchWord
+
+    await billpage.search(searchTerm)
+
+    const queryFilter = await billpage.queryFilter
+
+    await expect(queryFilter).toContainText("query:")
+    await expect(queryFilter).toContainText(searchTerm)
+  })
+
+  test("should click the bill and render to new page", async ({ page }) => {
+    // Perform a search and check the first bill on a random page
+    const billpage = new BillPage(page)
+    await billpage.goto()
+
+    const firstBillLink = await billpage.firstBill.evaluate(
+      (link: HTMLAnchorElement) => link.href
+    )
+    await billpage.clickFirstBill()
+    await expect(page.url()).toBe(firstBillLink)
+  })
+
+  test("should show no results found", async ({ page }) => {
+    // Test to ensure the application handles no results found cases
+    const searchTerm = "nonexistentsearchterm12345"
+    const billpage = new BillPage(page)
+    await billpage.goto()
+
+    billpage.search(searchTerm)
+
+    const noResultsText = await page.getByText("Looks Pretty Empty Here")
+    const noResultsImg = page.getByAltText("No Results")
+    const resultCounts = await billpage.resultCount
+
+    await expect(noResultsText).toBeVisible()
+    await expect(noResultsImg).toBeVisible()
+    await expect(resultCounts).toBeVisible()
+  })
+})
+
+// Define an interface for sorting test configurations
+interface SortingTest {
+  option: string
+  attribute: string
+  order: "asc" | "desc"
+  type:
+    | "relevance"
+    | "testimonyCount"
+    | "cosponsorCount"
+    | "nextHearingDate"
+    | "recentTestimony"
+}
+
+// Array of sorting test configurations
+// Need to add test for sort by relevant
+const sortingTests: SortingTest[] = [
+  {
+    option: "Sort by Testimony Count",
+    attribute: "div.testimonyCount",
+    order: "desc",
+    type: "testimonyCount"
+  },
+  {
+    option: "Sort by Cosponsor Count",
+    attribute: "span.blurb",
+    order: "desc",
+    type: "cosponsorCount"
+  },
+  {
+    option: "Sort by Next Hearing Date",
+    attribute: "div.card-footer",
+    order: "asc",
+    type: "nextHearingDate"
+  },
+  {
+    option: "Sort by Most Recent Testimony",
+    attribute: "span.blurb.me-2",
+    order: "asc",
+    type: "recentTestimony"
+  }
+]
+
+test.describe("Sort Bills test", () => {
+  for (const { option, attribute, order, type } of sortingTests) {
+    test(`should sort bills by ${option}`, async ({ page }) => {
+      const billpage = new BillPage(page)
+      await billpage.goto()
+
+      await billpage.sort(option)
+
+      // Verify the sorting result
+      const billValues = []
+      const sortedBills = await billpage.bills
+      const billsCount = await sortedBills.count()
+
+      for (let i = 0; i < billsCount; i++) {
+        const value = await billpage.getAttributeValue(
+          sortedBills.nth(i) as Locator,
+          attribute,
+          type
+        )
+        billValues.push(value)
```

Comment:
You should be able to do

```
const bill = sortedbills.nth(i)
const value = await bill.getAttribute("name of attribute")
```

I do something similar here: https://github.com/codeforboston/maple/pull/1580/files#diff-defcf5d940ff8eed10dfa30e3aeab17a4efaf8f26c9aa5014654be88041c5ea7R18-R25

Reply:
My final draft PR changed how I was testing the sorts, I was experience a lot of flakiness with this method. I decided to focus on clicking the sort button, and verifying if selected value changed. It doesn't test the full functionality but I think it's the best option we have for e2e testing, ensuring that the tests are stable.

https://github.com/codeforboston/maple/pull/1580/files#diff-defcf5d940ff8eed10dfa30e3aeab17a4efaf8f26c9aa5014654be88041c5ea7R152-R166

@Mephistic I forgot to mention this during the meeting, but I think the filtering and sorting tests for Bills and Testimonies might be better tested using unit tests instead. I haven't looked into it that in depth (I know Typesense might make things more difficult / work differently from what I expect), but I think using mocks will make verifying sort and filters easier

Reply:
👍 Makes sense to me - what we're really looking for here with the E2E tests isn't to validate the filter/sorting itself - it's to make sure the page doesn't blow up when we try to filter/sort.

Thanks for looking into this!

Reply:
Thank you for the reviews! @seatuna @Mephistic 

The getAttribute function is indeed more straight forward! My custom getAttributeValue was for verifying the sorting result which we are only focus on the page load it right and I will take it away.

I will also take away the functionality part of test in the filtering test.

Reply:
For the filter checkbox, I think it might be beneficial to use a locator to dynamically select the first available checkbox item instead of hard-coding a specific one. This way, the test will remain robust even if the data changes.

Reply:
@seatuna I found that encodeURIComponent does not encode single quotes ('). We can manually replace them with %27 as shown below

```javascript
const encodedFilterLabel = encodeURIComponent(filterLabel).replace(/'/g, "%27");